processing/src/main/java/io/druid/segment/StringDimensionHandler.java

```diff
@@ -0,0 +1,1609 @@
+    public ValueMatcher makeIndexingValueMatcher(
+        final Comparable matchValue,
+        final IncrementalIndexStorageAdapter.EntryHolder holder,
+        final int dimIndex
+    )
+    {
+      final String value = STRING_TRANSFORMER.apply(matchValue);
+      final int encodedVal = getEncodedValue(value, false);
+      final boolean matchOnNull = Strings.isNullOrEmpty(value);
+      if (encodedVal < 0 && !matchOnNull) {
+        return new BooleanValueMatcher(false);
+      }
+
+      return new ValueMatcher()
+      {
+        @Override
+        public boolean matches()
+        {
+          Object[] dims = holder.getKey().getDims();
+          if (dimIndex >= dims.length) {
+            return matchOnNull;
+          }
+
+          int[] dimsInt = (int[]) dims[dimIndex];
+          if (dimsInt == null || dimsInt.length == 0) {
+            return matchOnNull;
+          }
+
+          for (int i = 0; i < dimsInt.length; i++) {
+            if (dimsInt[i] == encodedVal) {
+              return true;
+            }
+          }
+          return false;
+        }
+      };
+    }
+
+    @Override
+    public ValueMatcher makeIndexingValueMatcher(
+        final DruidPredicateFactory predicateFactory,
+        final IncrementalIndexStorageAdapter.EntryHolder holder,
+        final int dimIndex
+    )
+    {
+      final Predicate<String> predicate = predicateFactory.makeStringPredicate();
+      final boolean matchOnNull = predicate.apply(null);
+      return new ValueMatcher()
+      {
+        @Override
+        public boolean matches()
+        {
+          Object[] dims = holder.getKey().getDims();
+          if (dimIndex >= dims.length) {
+            return matchOnNull;
+          }
+
+          int[] dimsInt = (int[]) dims[dimIndex];
+          if (dimsInt == null || dimsInt.length == 0) {
+            return matchOnNull;
+          }
+
+          for (int i = 0; i < dimsInt.length; i++) {
+            String finalDimVal = getActualValue(dimsInt[i], false);
+            if (predicate.apply(finalDimVal)) {
+              return true;
+            }
+          }
+          return false;
+        }
+      };
+    }
+
+    private void updateSortedLookup()
+    {
+      if (sortedLookup == null) {
+        sortedLookup = dimLookup.sort();
+      }
+    }
+
+    private String getActualValue(int intermediateValue, boolean idSorted)
+    {
+      if (idSorted) {
+        updateSortedLookup();
+        return sortedLookup.getValueFromSortedId(intermediateValue);
+      } else {
+        return dimLookup.getValue(intermediateValue);
+
+      }
+    }
+
+    private int getEncodedValue(String fullValue, boolean idSorted)
+    {
+      int unsortedId = dimLookup.getId(fullValue);
+
+      if (idSorted) {
+        updateSortedLookup();
+        return sortedLookup.getSortedIdFromUnsortedId(unsortedId);
+      } else {
+        return unsortedId;
+      }
+    }
+  }
+
+
+  public class StringDimensionMerger implements DimensionMergerV9
```

Comment:
It looks like there's a lot of code copied/pasted between here and StringDimensionMergerLegacy. Is it possible to share that? Maybe one extends the other (like the IndexMergers)? Or maybe they both call into some shared static functions? Or maybe the Legacy one has a V9 one as a member variable, and delegates to it for various ops?